src/workerd/io/worker.c++

```diff
@@ -893,59 +893,67 @@ struct Worker::Script::Impl {
       bool isException = false;
     };
 
-    modules.setDynamicImportCallback([](jsg::Lock& js, auto handler) mutable {
+    using Handler = kj::Function<jsg::Value()>;
+
+    static auto constexpr handleDynamicImport =
+        [](kj::Own<const Worker> worker,
+           Handler handler,
+           kj::Maybe<jsg::Ref<jsg::AsyncContextFrame>> asyncContext) ->
+               kj::Promise<DynamicImportResult> {
+      co_await kj::evalLater([] {});
+      auto asyncLock = co_await worker->takeAsyncLockWithoutRequest(nullptr);
```

Comment:
Non-blocking nit: It makes me a bit queasy to hold locks in the coroutine, but I guess that's what we gotta do. I think I would have a slight preference for moving some of the following logic into a regular function, though, perhaps everything after setting up the scopes.

Reply:
Yeah, I'm not entirely happy with it either but I'm not sure there's a better option. I don't want to bog down on it too much tho so I'm going to go ahead and get this landed then circle back around to the general problem of needing to hold locks within coroutines in general.